### PR TITLE
#1687 encode RegistryTypeEnum in HandlerType bits

### LINKED ISSUE #1687: Encode the auto_registry::RegistryTypeEnum directly in HandlerType bits
**What Needs to be Done?**

We can reduce a lot of code that makes assumptions and has to know based on context which handler type is being called by encoding the enum type bits directly in the handler. This is related to the bug #1685 because they get out of alignment.

### PR DESCRIPTION
**Pull request by jstrzebonski**
Fixes #1687

### CODE REVIEW

File: src/vt/registry/auto/auto_registry_type_enum.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
+/*
+//@HEADER
+// *****************************************************************************
+//
+//                            auto_registry_common.h
```
Comment (jstrzebonski) [inline src/vt/registry/auto/auto_registry_type_enum.h:5]: Wrong file name.

Comment (jstrzebonski) [reply src/vt/registry/auto/auto_registry_type_enum.h]: Done.

File: src/vt/handler/handler.cc [hunk 2]
```diff
@@ -47,7 +47,8 @@
 namespace vt {
 
 /*static*/ HandlerType HandlerManager::makeHandler(
-  bool is_auto, bool is_functor, HandlerIdentifierType id, bool is_objgroup,
+  bool is_auto, bool is_functor, HandlerIdentifierType id,
+  RegistryTypeEnum const registry_type, bool is_objgroup,
```
Comment (lifflander) [inline src/vt/handler/handler.cc:51]: Can we remove the two flags `is_objgroup` and `is_member` with the handler enum type encoded in the bits?

Comment (jstrzebonski) [reply src/vt/handler/handler.cc:51]: Yes, I think so.

I think I could even remove bits that encodes that values (`eHandlerBits::ObjGroup` and `Member`) as currently they're strictly bind with reg type. And for `isHandlerMember()` and `isHandlerObjGroup()`, I could just check reg type and return based on that, something like that:

```cpp
/*static*/ bool HandlerManager::isHandlerObjGroup(HandlerType han) {
  auto const reg_type = getHandlerRegistryType(han);
  auto const isObjGroup = reg_type == RegistryTypeEnum::RegObjGroup;
  return isObjGroup;
}

/*static*/ bool HandlerManager::isHandlerMember(HandlerType han) {
  auto const reg_type = getHandlerRegistryType(han);
  auto const isMember = reg_type ==  RegistryTypeEnum::RegVrtCollectionMember;
  return isMember;
}
```

Comment (lifflander) [reply src/vt/handler/handler.cc:51]: I think this would reduce the complexity even more then!

Comment (jstrzebonski) [reply src/vt/handler/handler.cc]: Done.

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -289,7 +289,8 @@ if(CODE_COVERAGE AND CMAKE_CXX_COMPILER_ID MATCHES "GNU|Clang")
   message(STATUS "VT: enabling code coverage generation")
   # Add required flags (GCC & LLVM/Clang)
   target_compile_options(${VIRTUAL_TRANSPORT_LIBRARY} PUBLIC
-    -Og
+    -O0        # no optimization
```
Comment (lifflander) [inline src/CMakeLists.txt:292]: Why does this change things back?

Comment (jstrzebonski) [reply src/CMakeLists.txt]: Done.

**Review by lifflander — APPROVED** [review]
Looks good!

**Review by PhilMiller — APPROVED** [review]

**Pull request merged**